storage/src/mmr/bitmap.rs

```diff
@@ -0,0 +1,290 @@
+    /// more typical balanced Merkle tree.
+    mmr: Mmr<H>,
+}
+
+impl<H: CHasher> Default for Bitmap<H> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<H: CHasher> Bitmap<H> {
+    const CHUNK_USIZE: usize = H::Digest::SERIALIZED_LEN;
+    const CHUNK_SIZE: u64 = H::Digest::SERIALIZED_LEN as u64;
+
+    /// Return a new empty bitmap.
+    pub fn new() -> Self {
+        Bitmap {
+            bitmap: vec![0u8; H::Digest::SERIALIZED_LEN],
+            next_bit: 0,
+            mmr: Mmr::new(),
+        }
+    }
+
+    /// Return the number of bits currently stored in the bitmap.
+    pub fn bit_count(&self) -> u64 {
+        self.bitmap.len() as u64 * 8 - (Self::CHUNK_SIZE * 8 - self.next_bit)
+    }
+
+    /// Commit the last chunk of the bitmap to the Merkle tree and initialize the next chunk.
+    fn commit_last_chunk(&mut self, hasher: &mut H) {
+        let len = self.bitmap.len();
+        let chunk = H::Digest::try_from(&self.bitmap[len - Self::CHUNK_USIZE..len]).unwrap();
+        self.mmr.add(hasher, &chunk);
+        self.next_bit = 0;
+        self.bitmap.extend(vec![0u8; Self::CHUNK_USIZE]);
+    }
+
+    /// Efficiently add a digest-sized chunk of bits to the bitmap. Assumes we are at a chunk
+    /// boundary (that is, `self.next_bit` is 0) and panics otherwise.
+    pub fn append_chunk_unchecked(&mut self, hasher: &mut H, chunk: &H::Digest) {
+        if self.next_bit != 0 {
+            panic!("cannot add chunk when not chunk aligned");
+        }
+
+        let len = self.bitmap.len();
+        let _ = &self.bitmap[len - Self::CHUNK_USIZE..len].copy_from_slice(chunk.as_ref());
+        self.commit_last_chunk(hasher);
+    }
+
+    /// Efficiently add a byte's worth of bits to the bitmap. Assumes self.next_bit is currently
+    /// byte aligned, and panics otherwise.
+    pub fn append_byte_unchecked(&mut self, hasher: &mut H, byte: u8) {
+        if self.next_bit % 8 != 0 {
+            panic!("cannot add byte when not byte aligned");
+        }
+
+        let len = self.bitmap.len() - Self::CHUNK_USIZE;
+        self.bitmap[len + self.next_bit as usize / 8] = byte;
+        self.next_bit += 8;
+        if self.next_bit == Self::CHUNK_SIZE * 8 {
+            self.commit_last_chunk(hasher);
+        }
+    }
+
+    /// Add a single bit to the bitmap.
+    pub fn append(&mut self, hasher: &mut H, bit: bool) {
+        let len = self.bitmap.len() - Self::CHUNK_USIZE;
+        if bit {
+            self.bitmap[len + self.next_bit as usize / 8] |= (1 << (self.next_bit % 8)) as u8;
+        }
+        self.next_bit += 1;
+        if self.next_bit == Self::CHUNK_SIZE * 8 {
+            self.commit_last_chunk(hasher);
+        }
+    }
+
+    /// Get the value of a bit.
+    pub fn get_bit(&self, bit_offset: u64) -> bool {
+        let byte_offset = bit_offset as usize / 8;
+        self.bitmap[byte_offset] & (1 << (bit_offset % 8)) != 0
+    }
```

Comment:
You might want to assert `bit_offset < self.bit_count()` to avoid reading from a byte that does technically exist in the `bitmap` but has never been formally set

Reply:
added.